dashboard/test/ui/features/applab/scenarios.feature

```diff
@@ -36,6 +36,17 @@ Feature: App Lab Scenarios
     Then element "#testButton1" contains text "Jelly"
     Then element "#testButton2" contains text "Jelly"
 
+  Scenario: Text is preserved when reading and setting newlines in textarea
+    Given I switch to design mode
+    And I drag a TEXT_AREA into the app
+    Then I switch to code mode
+    And I switch to text mode
+    And I press keys "setText('text_area1', 'Line 1\\nLine 2\\n\\nLine3');\n" for element ".ace_text-input"
+    And I press keys "for (var i = 0; i < 100; i++) { setText('text_area1', getText('text_area1')); }" for element ".ace_text-input"
+    When I press "runButton"
+    And I wait until element "#divApplab > .screen > div#text_area1" is visible
+    Then element "div#text_area1" has html "Line 1<div>Line 2</div><div><br></div><div>Line3</div>"
+
```

Comment:
I'll go ahead and add the unit test.  You are correct that I added the UI
test because I'm concerned about cross-browser behavior here.

Also realized this morning that I'm not cleaning HTML tags anymore, so I'll
test for that too.
On Oct 15, 2015 10:38 PM, "Bjvanminnen" <notifications@github.com> wrote:

> In dashboard/test/ui/features/applab/scenarios.feature
> <https://github.com/code-dot-org/code-dot-org/pull/4607#discussion_r42210063>
> :
>
> > @@ -36,6 +36,17 @@ Feature: App Lab Scenarios
> >      Then element "#testButton1" contains text "Jelly"
> >      Then element "#testButton2" contains text "Jelly"
> >
> > +  Scenario: Text is preserved when reading and setting newlines in textarea
> > +    Given I switch to design mode
> > +    And I drag a TEXT_AREA into the app
> > +    Then I switch to code mode
> > +    And I switch to text mode
> > +    And I press keys "setText('text_area1', 'Line 1\\nLine 2\\n\\nLine3');\n" for element ".ace_text-input"
> > +    And I press keys "for (var i = 0; i < 100; i++) { setText('text_area1', getText('text_area1')); }" for element ".ace_text-input"
> > +    When I press "runButton"
> > +    And I wait until element "#divApplab > .screen > div#text_area1" is visible
> > +    Then element "div#text_area1" has html "Line 1<div>Line 2</div><div><br></div><div>Line3</div>"
> > +
>
> I guess benefit of doing this in a UI test is we get cross browser. I
> think also adding a quick mochaTest (that would catch some stuff sooner)
> would also have some value. If doing so ends up being not super easy
> though, it's probably ok to just depend on UI test coverage.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/4607/files#r42210063>.
>